cfd_protocol/src/lib.rs

```diff
@@ -1,31 +1,154 @@
+use anyhow::Result;
+use bdk::bitcoin::Txid;
+use bdk::SignOptions;
+use bdk::{
+    bitcoin::{
+        self,
+        hashes::hex::ToHex,
+        util::bip143::SigHashCache,
+        util::psbt::Global,
+        util::{bip32::ExtendedPrivKey, psbt::PartiallySignedTransaction},
+        Address, Amount, Network, OutPoint, PrivateKey, PublicKey, Script, SigHash, SigHashType,
+        Transaction, TxIn, TxOut,
+    },
+    descriptor::Descriptor,
+    miniscript::{descriptor::Wsh, DescriptorTrait},
+    wallet::AddressIndex,
+};
+
+use rand::RngCore;
+use rand::{CryptoRng, SeedableRng};
+use rand_chacha::ChaChaRng;
+use secp256k1_zkp::{self, schnorrsig, Signature};
+use secp256k1_zkp::{bitcoin_hashes::sha256t_hash_newtype, SECP256K1};
+use secp256k1_zkp::{bitcoin_hashes::*, EcdsaAdaptorSignature};
+use secp256k1_zkp::{Secp256k1, SecretKey};
+use std::collections::HashMap;
+
+// TODO: Consider extension trait
+pub fn build_lock_psbt<B, D>(
+    wallet: &bdk::Wallet<B, D>,
+    caller_dlc_amount: Amount,
+) -> Result<PartiallySignedTransaction> {
+    todo!()
+}
+
+// TODO: Consider how this might change with off-chain protocol
+pub fn build_dlc_transactions(
+    maker: Party,
+    taker: Party,
+    refund_timelock: u32,
+    payouts: Vec<Payout>,
+    signing_sk: SecretKey,
+) -> Result<DlcTransactions> {
+    todo!()
+}
+
+fn dlc_descriptor(maker_pk: PublicKey, taker_pk: PublicKey) -> Descriptor<PublicKey> {
+    todo!()
+}
+
+fn spend_tx_sighash(
+    tx: &Transaction,
+    dlc_descriptor: &Descriptor<PublicKey>,
+    dlc_amount: Amount,
+) -> SigHash {
+    todo!()
+}
+
+pub fn finalize_spend_transaction(
+    tx: Transaction,
+    dlc_descriptor: Descriptor<PublicKey>,
+    (maker_pk, maker_sig): (PublicKey, Signature),
+    (taker_pk, taker_sig): (PublicKey, Signature),
+) -> Transaction {
+    todo!()
+}
+
+// NOTE: We have decided to not offer any verification utility because
+// the APIs would be incredibly thin
+
+pub struct Party {
+    pub lock_psbt: PartiallySignedTransaction,
+    pub dlc_pk: PublicKey,
+    pub dlc_amount: Amount,
+    pub address: Address,
+}
```

Comment:
Creating one of these requires funding the psbt correct?

Maybe a extension trait above should already return you a `Party` struct?

Reply:
The wallet could fund the psbt and automatically fill in the address!

Reply:
Great idea! The only thing that's less obvious is who should generate the corresponding `dlc_sk`? I'm leaning towards generating before and passing in `dlc_pk` to the trait method.